File: flink-cyber/metron-parser-chain/parser-chains-config-service/frontend/parser-chains-client/src/app/services/pipeline.service.ts
/*
 * Copyright 2020 - 2022 Cloudera. All Rights Reserved.
 *
 * This file is licensed under the Apache License Version 2.0 (the "License"). You may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. Refer to the License for the specific permissions and
 * limitations governing your use of the file.
 */

import {HttpClient, HttpParams} from '@angular/common/http';
import {Injectable} from '@angular/core';

import {map} from 'rxjs/operators';
import {getHttpParams} from '../shared/service.utils';

@Injectable({
    providedIn: 'root'
})
export class PipelineService {

    static readonly BASE_URL = '/api/v1/pipeline';

    constructor(
        private _http: HttpClient,
    ) {
    }

    public getPipelines() {
        const httpParams: HttpParams = getHttpParams(null);
        return this._http.get<string[]>(PipelineService.BASE_URL, {params: httpParams})
            .pipe(map(strArr => strArr.map(pipelineName => {
                return pipelineName
            })));
    }

    public createPipeline(pipelineName: string) {
        const httpParams: HttpParams = getHttpParams(null);

        return this._http.post<string[]>(PipelineService.BASE_URL + "/" + pipelineName,
            null,{params: httpParams})
            .pipe(map(strArr => strArr.map(pName => {
                return pName
            })));
    }

    public renamePipeline(pipelineName: string, newPipelineName: string) {
        let httpParams: HttpParams = getHttpParams(null);
        httpParams = httpParams.set('newName', newPipelineName)

        return this._http.put<string[]>(PipelineService.BASE_URL + "/" + pipelineName,
            null,{params: httpParams})
            .pipe(map(strArr => strArr.map(pName => {
                return pName
            })));
    }

    public deletePipeline(pipelineName: string) {
        const httpParams: HttpParams = getHttpParams(null);

        return this._http.delete<string[]>(PipelineService.BASE_URL + "/" + pipelineName, {params: httpParams})
            .pipe(map(strArr => strArr.map(pName => {
                return pName
            })));
    }
}

File: flink-cyber/metron-parser-chain/parser-chains-config-service/src/main/java/com/cloudera/parserchains/queryservice/service/PipelineService.java
/*
 * Copyright 2020 - 2022 Cloudera. All Rights Reserved.
 *
 * This file is licensed under the Apache License Version 2.0 (the "License"). You may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. Refer to the License for the specific permissions and
 * limitations governing your use of the file.
 */

package com.cloudera.parserchains.queryservice.service;

import com.cloudera.parserchains.queryservice.config.AppProperties;
import com.cloudera.parserchains.queryservice.model.exec.PipelineResult;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.flink.core.fs.FileStatus;
import org.apache.flink.core.fs.FileSystem;
import org.apache.flink.core.fs.Path;
import org.springframework.cache.CacheManager;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class PipelineService {

  private final AppProperties appProperties;
  private final CacheManager cacheManager;

  @Cacheable("pipelinePathMap")
  public Map<String, PipelineResult> findAll() throws IOException {
    final Path pipelineRootPath = getPipelineRootPath();

    final FileSystem fileSystem = pipelineRootPath.getFileSystem();
    if (!fileSystem.exists(pipelineRootPath)) {
      return null;
    }
    final FileStatus[] statusList = fileSystem.listStatus(pipelineRootPath);
    if (statusList == null) {
      return null;
    }

    final Map<String, PipelineResult> pipelineMap = new HashMap<>();
    for (FileStatus fileStatus : statusList) {
      if (fileStatus.isDir()) {
        final Path pipelinePath = fileStatus.getPath();
        if (isValidPipeline(pipelinePath, fileSystem)) {
          final String name = pipelinePath.getName();
          final PipelineResult pipeline = PipelineResult.builder()
              .name(name)
              .path(pipelinePath)
              .build();
          pipelineMap.put(name, pipeline);
        }
      }
    }
    return pipelineMap;
  }

  public PipelineResult createPipeline(String pipelineName) throws IOException {
    final Path pipelineRootPath = getPipelineRootPath();
    final FileSystem fileSystem = pipelineRootPath.getFileSystem();
    if (!fileSystem.exists(pipelineRootPath)) {
      return null;
    }

    final Path pipelinePath = new Path(pipelineRootPath.getPath(), pipelineName);
    final Path fullValidPath = new Path(pipelinePath, "parse/chains");

    if (fileSystem.mkdirs(fullValidPath)) {
      evictPipelineCache();
      return PipelineResult.builder()
          .name(pipelineName)
          .path(fullValidPath)
          .build();
    }
    return null;
  }

  public PipelineResult renamePipeline(String pipelineName, String newName) throws IOException {
    final Path pipelineRootPath = getPipelineRootPath();
    final FileSystem fileSystem = pipelineRootPath.getFileSystem();
    if (!fileSystem.exists(pipelineRootPath)) {
      return null;
    }

    final Path originalPath = new Path(pipelineRootPath, pipelineName);
    if (!isValidPipeline(originalPath, fileSystem)) {
      throw new RuntimeException("Provided original pipeline name doesn't correspond to a valid pipeline!");
    }
    final Path newPath = new Path(pipelineRootPath, newName);

    if (fileSystem.exists(newPath)) {
      throw new RuntimeException("New pipeline name is already occupied!");
    }

    if (fileSystem.rename(originalPath, newPath)) {
      evictPipelineCache();
      return PipelineResult.builder()
          .name(newName)
          .path(newPath)
          .build();
    }

    return null;
  }

  public boolean deletePipeline(String pipelineName) throws IOException {
    final Path pipelineRootPath = getPipelineRootPath();
    final FileSystem fileSystem = pipelineRootPath.getFileSystem();
    if (!fileSystem.exists(pipelineRootPath)) {
      return false;
    }

    final Path pipelinePath = new Path(pipelineRootPath, pipelineName);
    if (!isValidPipeline(pipelinePath, fileSystem)) {
      throw new RuntimeException("Provided pipeline name doesn't correspond to a valid pipeline!");
    }
    if (fileSystem.delete(pipelinePath, true)){
      evictPipelineCache();
      return true;
    }
    return false;
  }

  private boolean isValidPipeline(Path originalPath, FileSystem fileSystem) throws IOException {
    final Path fullPath = new Path(originalPath, "parse/chains");
    return fileSystem.exists(fullPath);
  }

  private Path getPipelineRootPath() {
    String pipelinePathStr = appProperties.getPipelinesPath();
    return new Path(pipelinePathStr);
  }

  private void evictPipelineCache() {
    cacheManager.getCache("pipelinePathMap").clear();
  }

}
